shopBuddy/shopBuddy/GlobalManager.m

```diff
@@ -0,0 +1,197 @@
+        
+    }
+    return self;
+}
+
+- (void)fetchPricesWithItem:(Item *)item fromStore: (NSString *)store completion:(void(^)(NSDictionary *prices, BOOL success))completion {
+    //TODO: Parse Dictionary and return Array of NSObject Price
+    NSString *const jobId = self.itemJobIdMap[item.name];
+    if (jobId == nil) {
+        [self _submitJob:item.name withStore:store withCompletion:^(NSString *job_id, NSError *error) {
+            if(!error) {
+                self.itemJobIdMap[item.name] = job_id;
+                [self _checkJobStatus:job_id withCompletion:^(NSDictionary *prices, BOOL success) {
+                    completion(prices,success);
+                }];
+            }
+        }];
+    } else {
+        [self _checkJobStatus:jobId withCompletion:^(NSDictionary *prices, BOOL success) {
+            completion(prices,success);
+        }];
+    }
+}
+
+- (void)_checkJobStatus: (NSString *)job_id withCompletion:(void(^)(NSDictionary *prices, BOOL success))completion {
+    if (self.completeJobs[job_id]!=nil) {
+        completion(self.completeJobs[job_id],YES);
+        //TODO: check if stale
+        return;
+    }
+
+    if ([self.outstandingJobs containsObject:job_id]) {
+        // Time window passed but last sync is not ready
+        __weak __typeof(self) weakSelf = self;
+        dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
+        const dispatch_time_t timeoutTime = dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC);
+        dispatch_after(timeoutTime, queue, ^{
+            [weakSelf _checkJobStatus:job_id withCompletion:^(NSDictionary *prices, BOOL success) {
+                if(prices!=nil) {
+                    completion(prices,success);
+                }
+            }];
+        });
+        return;
+    }
+
+    // Either is a new job or is retry for an incomplete job
+    __weak __typeof(self) weakSelf = self;
+    dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
+    dispatch_async(queue, ^{
+        [self.outstandingJobs addObject:job_id];
+        [weakSelf requestJobStatus:job_id withCompletion:^(BOOL jobIsFinished, NSError *error) {
+            [weakSelf _jobStatusCallback:job_id finished:jobIsFinished withCompletion:^(NSDictionary *prices, BOOL success) {
+                completion(prices,success);
+            }];
+        }];
+    });
+}
+
+- (void)_jobStatusCallback:(NSString *)jobId finished: (BOOL)finished withCompletion:(void(^)(NSDictionary *prices, BOOL success))completion {
+    if (finished) {
+        __weak __typeof(self) weakSelf = self;
+        [self downloadJobResults:jobId withCompletion:^(NSDictionary *results, NSError *error) {
+            __strong __typeof(weakSelf) strongSelf = weakSelf;
+            if (strongSelf != nil) {
+                if(!error) {
```

Comment:
Move the `if(!error)` up so you don't need to instantiate `strongSelf` if `false`